### PR TITLE
Preferences: Redux

### PR DESCRIPTION
**Pull request by itsmeow**
## About The Pull Request

Using TGUI preferences as a basis, rewrites the entire preferences backend and frontend. Preferences are now datumized and can be added by simply creating a datum and a typescript entry for the preference. Character preferences require a new DB column as well, however.

Our preference serialization system is basically a complete rewrite from TG's, and is now far more robust. We could even theoretically add back savefile support for local users, since serialization is now almost entirely abstracted from the serialization method.

All preference writes are cached into a subsystem that fires every 5 seconds, limiting successive writes increasing DB traffic. All preferences, datumized or undatumized, now have a "dirty" flag, that ensures unmodified preferences don't rewrite to the database, reducing effective database traffic by 95% for most preference edits.

Preference reads are also stored into a local cache object, the same on that tracks what preferences are dirty.

I've also added the entire loadouts system back to the menu, and rewritten the preference access API to more strictly differentiate between character and player preferences.

Many components have been restyled and many will be reworked to have better UX.

### DOWNSTREAMS

**Pre-requisite PRs for this PR:**
- #9229

**Fix PRs for this PR**
- #9650
- #9661 (semi-important)
- #9662
- #9674

### BeeStation Changes (So Far)

- Rewrite of preference serialization for database
- Database write queueing system
- Undatumized preference dirty flagging
- Datumized preference dirty flagging
- Entire loadouts system
- Datumized antagonist/role preferences rather than tying to dynamic and using TS files.
- Apids, IPCs, Oozelings - new species perks, etc
- Code organization: Reorganized all the files in the prefs module to make sense
- Quirks: Some different and new icons for our quirks
- Improved mannequin rendering: complete BYOND Bug workaround (+ I made the fuckin report https://www.byond.com/forum/post/2873835)
- No more legacy_toggles (yeah fuck you)
- Distinct read_player_preference and read_character_preference
- update_preference is designed to work outside the menu now
- Better verbs organization
- Slightly restyled Occupations selection
- More previews / fixed species feature previews (moth antannae, lizard parts etc)
- Added search to role/antag preferences
- Improved character profiles data caching/updating
- Improved SQL error handling
- Improved fetch error handling
- Re-added roundstart_no_hard_check
- Altered character profile caching
- Better flex layout for Character preferences menu buttons (multiline support)
- Better character preview screen registration/setup
- Fixed page being invalid when switching between character/game prefs.
- Added search bar to MainFeatures
- Increased MainFeature preview sizes, added Layout theming to popups
- Filled whitespace in character customization
- Larger Alternate Names button for better UX.
- Added second preview rotation button
- Added preset palettes for eye, skin, hair, etc colors.
- Supplemental features can now display at the bottom of the popup window.
- "None"/"Bald"/"Shaved" hair types are put first in the style order.
- Game preferences are searchable
- Game preferences have better multiline support
- Game preferences now have subcategories and have been reorganized to make more sense.
- Per-character antag prefs
- Being banned from an antag does not disable the preference button for it
- Enable/Disable all antag prefs button for all categories
- Scales pref fake popups to match their content's size
- Improved gender popup, with outside tracking and better styling
- drop shadow on fake popups
- Changed game preferences layouting to have higher density / flow

TGUI preferences basis + fixes

- https://github.com/tgstation/tgstation/pull/61313
- https://github.com/tgstation/tgstation/pull/61492
- https://github.com/tgstation/tgstation/pull/61519
- https://github.com/tgstation/tgstation/pull/61542
- https://github.com/tgstation/tgstation/pull/61586
- https://github.com/tgstation/tgstation/pull/61764
- https://github.com/tgstation/tgstation/pull/61821
- https://github.com/tgstation/tgstation/pull/61942
- https://github.com/tgstation/tgstation/pull/61957
- https://github.com/tgstation/tgstation/pull/61973
- https://github.com/tgstation/tgstation/pull/62218
- https://github.com/tgstation/tgstation/pull/62394 (partial)
- https://github.com/tgstation/tgstation/pull/62653
- https://github.com/tgstation/tgstation/pull/62888
- https://github.com/tgstation/tgstation/pull/63128
- https://github.com/tgstation/tgstation/pull/63169
- https://github.com/tgstation/tgstation/pull/63200
- https://github.com/tgstation/tgstation/pull/63221
- https://github.com/tgstation/tgstation/pull/63225 (partial)
- https://github.com/tgstation/tgstation/pull/63370
- https://github.com/tgstation/tgstation/pull/63554
- https://github.com/tgstation/tgstation/pull/64610
- https://github.com/tgstation/tgstation/pull/65140
- https://github.com/tgstation/tgstation/pull/65645
- https://github.com/tgstation/tgstation/pull/70577
- https://github.com/tgstation/tgstation/pull/74384
- https://github.com/tgstation/tgstation/pull/75208

Other TGUI

- https://github.com/tgstation/tgstation/pull/75164
- https://github.com/tgstation/tgstation/pull/75367

Feature pre-reqs

- https://github.com/tgstation/tgstation/pull/46952 (partial)
- https://github.com/tgstation/tgstation/pull/47670 (partial)
- https://github.com/tgstation/tgstation/pull/55560
- https://github.com/tgstation/tgstation/pull/55948

Assets stuff

- https://github.com/tgstation/tgstation/pull/62845
- https://github.com/tgstation/tgstation/pull/66549

Loadouts improvements

- https://github.com/Royale-Station-13/Royale-Station-13/pull/37

## Why It's Good For The Game

Preferences is one of the most used menus, and also one of the worst in design and function. Having it on TGUI will allow more rapid iteration. Same with datumizing prefs, currently you had to write a ton of boilerplate to just add one preference.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

image
image
image
image
image
image
image
image

</details>

## Changelog
:cl: itsmeow, Mothblocks, Absolucy, Rukofamicom, Ghommie, JohnFulpWillard, Watermelon914, PeterMorrison1, Y0SH1M4S73R, AnturK, GoldenAlpharex, Seris02, tralezab, AndrewL97, MrMelbert, Tastyfish, LemonInTheDark, vinylspiders, kriskog, Couls, MrStonedOne, jlsnow301, actioninja, Synclt21, KoJIT2009
refactor: Rewrote the entire preferences UI.
refactor: Rewrote the entire preferences backend and serialization system to reduce database usage, datumized the addition of preferences.
server: Flattened some preference database schema stuff, migration is required.
tweak: Loadouts now preview the items you equip on your mob, and will replace existing items in the preview window to ensure they preview correctly.
fix: Deadminned admins now retain their access to Donator loadout items.
add: You can now purchase multiple antag tokens from loadouts.
add: You can now bind up to 3 keybinds to one overall action.
add: You can now select specific preferences to be randomized on join rather than globally randomizing your body. You can also make specific preferences only randomize if you are an antagonist.
tweak: PDA Theme can now be customized per-character.
tweak: Failed job selection now defaults to "Join as assistant" instead of "Be random job"
tweak: TGUI Dropdowns now have improved UX experience, and exist outside of the object they are within, instead existing above other UI elements. This means they cannot cause windows to scroll by overflowing, for example.
tweak: You can now select preferences to be randomized only when you are an antagonist
tweak: Per-character antagonist preferences can now be toggled globally, with per-character prefs only affecting if it's set to off.
/:cl:

### CODE REVIEW

**Comment by CydiaLamiales** [comment]
Massive pog

**Comment by itsmeow** [comment]
I literally just badly fixed merge conflicts and committed my changes, don't congratulate me yet dammit

**Comment by imsynthex** [comment]
if it compiles, it's practically done

**Comment by imsynthex** [comment]
eh, good enough

**Comment by Rukofamicom** [comment]
Port looks great, congrats

**Comment by itsmeow** [comment]
stoop congratulating me it's nowhere near usable 😭😭

**Comment by CydiaLamiales** [comment]
<insert Evangelion 'Congratulations' gif here>

**Comment by Mothblocks** [comment]
i do believe theres a name youre forgetting in the changelog

**Comment by itsmeow** [comment]
of course I'll add it all when I'm done, thank you for the code

**Comment by itsmeow** [comment]
of course I'll add it all when I'm done, thank you for the code

File: code/game/machinery/telecomms/broadcasting.dm [hunk 1]
```diff
@@ -171,11 +171,6 @@
 	// From the list of radios, find all mobs who can hear those.
 	var/list/receive = get_mobs_in_radio_ranges(radios)
 
-	// Cut out mobs with clients who are admins and have radio chatter disabled.
-	for(var/mob/R in receive)
-		if (R.client && R.client.holder && !(R.client.prefs.chat_toggles & CHAT_RADIO))
-			receive -= R
```
Comment (itsmeow) [inline code/game/machinery/telecomms/broadcasting.dm:177]: ~~possibly bad removal~~ re-added

**Comment by itsmeow** [comment]
Hey remember how you were all congratulating me??


it does actually sort of work now. you can like, run it. that is, if you update your local database schema. but it's very cool

**Comment by itsmeow** [comment]
This PR is now at a stage where it is code reviewable.

File: code/modules/antagonists/nukeop/nukeop.dm [hunk 2]
```diff
@@ -20,9 +24,6 @@
 	var/datum/atom_hud/antag/opshud = GLOB.huds[ANTAG_HUD_OPS]
 	opshud.leave_hud(M)
 	set_antag_hud(M, null)
-
-/datum/antagonist/nukeop/apply_innate_effects(mob/living/mob_override)
```
Comment (itsmeow) [inline code/modules/antagonists/nukeop/nukeop.dm]: BAD

**Comment by WrillWasTaken** [comment]
Don't forget [wiki changes](https://wiki.beestation13.com/view/Starter_guide) [here,](https://wiki.beestation13.com/view/File:Character_settings.PNG) [here,](https://wiki.beestation13.com/view/File:New_Occupation_screen.JPG) and [here.](https://wiki.beestation13.com/view/File:Game_prefences_Screen.PNG)

File: SQL/database_changelog.txt [hunk 3]
```diff
@@ -1,15 +1,541 @@
...
+DO NOT RUN WITHOUT TAKING A FULL BACKUP.
+DO NOT RUN MORE THAN ONCE.
+DO NOT RUN COLUMN REMOVALS UNTIL DATA IS VERIFIED.
+- Allows nulls for `ss13_characters` columns (required for partial INSERT INTO ON DUPLICATE KEY UPDATE)
+- Alters some datatypes to fit new data.
+- Turns `ss13_characters`.`features` and `ss13_characters`.`custom_names` JSON objects into individual columns, removes the original columns.
+- Updates `ss13_preferences`.`preference_tag` to text.
+- Updates various `ss13_preferences` values to new ones.
+- Flips `ss13_preferences`.`key_bindings` JSON structure.
+
```
Comment (Crossedfall) [inline SQL/database_changelog.txt]: This is too much for a simple CL entry. Put it in its own SQL file similar to the other ones. Call it something like `prefs_migration_8776.sql`

**Comment by EvilDragonfiend** [comment]
image
image

Box is slightly small. not that an issue though...

-----

When your pref window is open, mouse cursor is flickering. I feel like it's keeping loading something? or is it refreshing?

-----

image

I think a character slot relocation would be nice (as long as slots are open)
Once their Byond membership is gone, it should pop up a window like `Your character slots are now limited due to the expiration of your Byond membership. Please select X characters to play. Non-selected characters will be locked until further membership.`

-------

image

pref window should be slightly bigger. a scrollbar for 5px is ugly

-------

* Loadout might need a preview button
* I think purchased items should specify which part(category) it belongs to.

-----

image

I hardly understand the color relation when it comes with C flag.

---------

image


This is not well readible. Should be white-text or something.
or it should be light-red background at least.

**Comment by EvilDragonfiend** [comment]
purchase button might need a confirmation button

**Comment by itsmeow** [comment]
> I hardly understand the color relation when it comes with C flag.

Green: On, you will get this antag
Grey: Off due to C button or banned/not enough playtime
Red: Disabled intentionally.

**Comment by itsmeow** [comment]
> pref window should be slightly bigger. a scrollbar for 5px is ugly

This seems to be a you problem, because it's fine for me.

**Comment by Crossedfall** [comment]
The scrollwheel for the loadout tab isn't active by default. I assume this is because the outer scrollwheel has priority or the inner-box isn't active until on-click. It would be superb if we could change that.

**Comment by itsmeow** [comment]
> The scrollwheel for the loadout tab isn't active by default. I assume this is because the outer scrollwheel has priority or the inner-box isn't active until on-click. It would be superb if we could change that.

This is mainly because IE11 has absolute garbage scroll capturing, rather than being based on what scrollable box you are hovering, it's based on the currently active element (by clicking). I am not sure how to fix this, but I'll look further into it.

**Comment by Miliviu** [comment]
just came back to congrat you for this, awesome port Kat!

**Comment by bluezorua** [comment]
you can't seem to recolour IPC antennas? they seem to be stuck black, atleast on the char editor

**Review by PowerfulBacon — APPROVED** [review]

**Review by Crossedfall — APPROVED** [review]

**Pull request merged**